### ISSUE TITLE
Use query parameters rather than string building in Neo4jCsvPublisher

### ISSUE THREAD

**Issue by jonhehir** [amundsen-io/amundsen#513]
As noted in https://github.com/lyft/amundsendatabuilder/pull/291, the `Neo4jCsvPublisher` class is currently using a string building approach to build Cypher queries for mass data import. Unfortunately, it's reasonably easy to accidentally (or intentionally!) build invalid queries, which cause entire imports to fail. In the linked PR, for example, a description of `This is a test: \_.` causes an issue, as the `\_` sequence is interepreted as an invalid escape sequence as opposed to the literal sequence of characters `\_`.

While this specific issue is patched up by escaping backslashes, I imagine other issues remain. For example, if for some strange reason you try to import a table whose description contains a null byte, the Neo4jCsvPublisher will trip over this as well. While this is perhaps not an especially common use case, it would be nice if the process were generally robust to a wide variety of string inputs and not easily broken.

## Expected Behavior or Use Case

Basically, the `Neo4jCsvPublisher` needs to be robust to various string inputs. 

## Service or Ingestion ETL

`Neo4jCsvPublisher`

## Possible Implementation

In my opinion, the cleanest and safest path to dealing with this would be to use [query parameters](https://neo4j.com/docs/cypher-manual/current/syntax/parameters/#cypher-parameters-create-node-with-properties) for these import statements. Query parameters are (thankfully) already in use on the user-facing side of things in `Neo4jProxy`.

**Comment by feng-tao** [comment]
https://github.com/lyft/amundsendatabuilder/pull/291 is merged. I assume this is fixed.

**Comment by feng-tao** [comment]
thanks @jonhehir

**Comment by feng-tao** [comment]
oh, actually we could still open it to keep track the long term request

**Comment by dorianj** [comment]
I looked into query builders for neo4j and didn't find a clear standard. There's https://github.com/emehrkay/Pypher (88 GH stars) and https://github.com/TwoBitAlchemist/NeoAlchemy (21 stars). Both are attempting to be general-purpose query-builders, meaning much of the code would be need to be rewritten to use their composition style. Does anyone know of any other such projects that may be more popular?

In my experience, using these sorts of query builders for anything other than trivial queries results in much pain, since you end up having to hit the string escape hatch anyhow, and now you're dealing with the worst of both worlds: both string templates for parts, and query builder for other parts. I think that's compounded by these query builders being young/immature -- they're more likely to be missing support for edge cases.

Given that, I think the best solution here is to pass around a context object that tracks arbitrary number of parameters, something like this:

```
class NamedParameters:
    """
    Class that tracks state of names and values of parameters, allowing creation of
    anonymous parameter names.
    """
    def __init__(self):
        self._params = dict()
        self._counter = 0

    def _param_name(self):
        self._counter += 1
        return "{}_{}".format(prefix, self._counter)

    def param(self, value, prefix='p'):
        """
        Returns a new anonymous param name, and stores the value. Pass in a meaningful
        prefix to make the query easier to debug from logs.
        """
        name = self._param_name()
        self._params[name] = value
        return name
```

Then in the calling code, instead `props.append('{id}.{key} = {val}'.format(id=identifier, key=k, val=v))`, we'd do something like this:
```
param_name = param_context.param(v)
props.append('{id}.{key} = {param_name}'.format(id=identifier, key=k, val=param_name))
```

(where `param_context` is passed around through the calltree for each query). It will be a bit confusing because there are two levels of templating, the `Template` and also the call-time params, but I think it would be OK.

I'm a bit wary of going out and doing this since the current unit test doesn't check for query correctness, only that the correct number of queries pass. So ideally I'd want to first add an integration test that tests correctness (maybe this is already present somewhere but I'm missing it?). I'm going to revisit this once I have a better handle on the databuilder code and feel like I can land it in a way that's unlikely to cause regressions.

**Comment by feng-tao** [comment]
https://github.com/amundsen-io/amundsendatabuilder/pull/382 is merged now, let us know if it resolves your issue.

**Comment by jonhehir** [comment]
Thanks, @feng-tao!
